BridgeSDK/BridgeAPI/SBBActivityManager.m

```diff
@@ -65,75 +67,99 @@ - (NSURLSessionDataTask *)getScheduledActivitiesForDaysAhead:(NSInteger)daysAhea
-                NSComparisonResult compareToStart = [activity.expiresOn compare:windowStartDate];
-                BOOL withinDaysBehind = ((compareToToday == NSOrderedAscending) || (compareToToday == NSOrderedSame)) && ((compareToStart == NSOrderedDescending) || (compareToStart == NSOrderedSame));
-                
-                return withinDaysBehind && activity.finishedOn == nil;
-            }];
-            
-            daysBehindTasks = [tasks.items filteredArrayUsingPredicate:predicate];
-        }
+        savedTasks = [self filterTasks:tasks.items forDaysAhead:kMaxAdvance andDaysBehind:kDaysToCache excludeStillValid:YES];
     }
 
     return [self.networkManager get:kSBBActivityAPI headers:headers parameters:@{@"daysAhead": @(daysAhead), @"offset": [[NSDate date] ISO8601OffsetString]} completion:^(NSURLSessionDataTask *task, id responseObject, NSError *error) {
-        tasks = [self.objectManager objectFromBridgeJSON:responseObject];
+        SBBResourceList *tasks = [self.objectManager objectFromBridgeJSON:responseObject];
         if (policy == SBBCachingPolicyFallBackToCached) {
             // we've either updated the cached tasks list from the server, or not, as the case may be;
             // in either case, we want to pass the cached tasks list to the completion handler.
-            tasks = (SBBResourceList *)[cacheManager cachedSingletonObjectOfType:@"ResourceList" createIfMissing:NO];
+            tasks = (SBBResourceList *)[cacheManager cachedObjectOfType:@"ResourceList" withId:@"ScheduledActivity" createIfMissing:NO];
             
             // ...ok, if we *did* update from the server though, we want to add back the ones left over from
             // the daysBehind window.
-            if (!error && daysBehindTasks.count) {
-                [tasks insertItems:daysBehindTasks atIndexes:[NSIndexSet indexSetWithIndexesInRange:NSMakeRange(0, daysBehindTasks.count)]];
+            if (!error && savedTasks.count) {
+                [tasks insertItems:savedTasks atIndexes:[NSIndexSet indexSetWithIndexesInRange:NSMakeRange(0, savedTasks.count)]];
                 
                 // ...and save it back to the cache for later.
                 [tasks saveToCoreDataCacheWithObjectManager:self.objectManager];
             }
         }
         if (completion) {
-            completion(tasks, error);
+            NSMutableArray *requestedTasks = [[self filterTasks:tasks.items forDaysAhead:daysAhead andDaysBehind:daysBehind excludeStillValid:NO] mutableCopy];
+            if ([self.objectManager conformsToProtocol:@protocol(SBBObjectManagerInternalProtocol)]) {
+                // in case object mapping has been set up
+                for (NSInteger i = 0; i < requestedTasks.count; ++i) {
+                    requestedTasks[i] = [((id<SBBObjectManagerInternalProtocol>)self.objectManager) mappedObjectForBridgeObject:requestedTasks[i]];
+                }
+            }
```

Comment:
can you pull this out as a block or a convenience method? Looks very similar to the code above.